### ISSUE TITLE
Omegaconf from_argparse - Provide some interoperability with default argument parser

### ISSUE THREAD

**Issue by georgepar** [omry/omegaconf#569]
**Is your feature request related to a problem? Please describe.**

OmegaConf.from_cli is good, but most frameworks / individuals already use argparse for parsing command line arguments.
Specifically I'm thinking about the use case of having a config file with some model training configuration and being able to override some properties using command line args for fast experimentation.

**Describe the solution you'd like**

In my code base I've come up with this hack

```python
class OmegaConfExtended(OmegaConf):
    @staticmethod
    def from_argparse(parser: argparse.ArgumentParser):
        dest_to_arg = {v.dest: k for k, v in parser._option_string_actions.items()}

        all_args = vars(parser.parse_args())
        provided_args = {}
        default_args = {}
        for k, v in all_args.items():
            if dest_to_arg[k] in sys.argv:
                provided_args[k] = v
            else:
                default_args[k] = v

        provided = OmegaConf.create(_nest(provided_args))
        defaults = OmegaConf.create(_nest(default_args))

        return provided, defaults
```

Basically, it takes an argument parser, parses the arguments and splits them in default and user-provided arguments.
Then I can do something like `config = OmegaConf.merge(default_cli_args, config_from_yaml, user_provided_cli_args)`.
(Merge precedence: user provided CLI args > Config file > default values for CLI args)

**Describe alternatives you've considered**
The alternative is to use the default `from_cli`.

**Additional context**

Off-course this solution makes a lot of assumptions (especially about nesting).

Specifically I do something like this for nested args:

```python
parser.add_argument("--hidden", dest="model.hidden", default=1233, type=int)
```
and then have recursive function that converts `{"model.hidden": 1233}` to `{"model": {"hidden": 1233}}`

Has anybody tried something similar / toyed with this idea?

**Comment by omry** [comment]
Hi,
I general OmegaConf offers really limited support for CLI integration and I am not interested in extending it.
Frameworks like Hydra are building much more powerful integrations.

from_cli is a very thin wrapper around [OmegaConf.from_dotlist](https://github.com/omry/omegaconf/blob/cab594e9b195019a2a5993ec77e755717f98d24a/omegaconf/omegaconf.py#L326):
```python
    @staticmethod
    def from_cli(args_list: Optional[List[str]] = None) -> DictConfig:
        if args_list is None:
            # Skip program name
            args_list = sys.argv[1:]
        return OmegaConf.from_dotlist(args_list)
```

You are welcome to build your own version using `OmegaConf.from_dotlist()`.

**Comment by omry** [comment]
Check the [docs](https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#from-a-dot-list) of from_dotlist.

**Comment by Jasha10** [comment]
@georgepar Are you familiar with https://hydra.cc? It uses OmegaConf as its backend and has very good support for reading multiple yaml files and overriding those files with command-line arguments.

**Comment by georgepar** [comment]
@Jasha10 Hi, thanks for the suggestion! Yes I've looked into hydra, but I ended up not using it because I was looking for something more barebones (i.e. not having to wrap your main with a decorator etc.), and I prefer to have one monolithic config file that can be passed around for easy reproduction.

Now I'm browsing their docs and maybe I will take another look at their compose API, which looks interesting.

Until then my little hack has actually been proven good enough for my usecase :)
